packages/dds/tree/src/test/simple-tree/objectNode.spec.ts

```diff
@@ -0,0 +1,125 @@
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+
+import { validateAssertionError } from "@fluidframework/test-runtime-utils/internal";
+
+import { SchemaFactory } from "../../simple-tree/index.js";
+
+import { hydrate } from "./utils.js";
+import { requireAssignableTo } from "../../util/index.js";
+
+const schemaFactory = new SchemaFactory("Test");
+
+describe("ObjectNode", () => {
+	describe("shadowing", () => {
+		it("optional shadowing builtin", () => {
+			class Schema extends schemaFactory.object("x", {
+				toString: schemaFactory.optional(schemaFactory.number),
+			}) {}
+			{
+				const n = hydrate(Schema, { toString: 1 });
+				assert.equal(n.toString, 1);
+				// @ts-expect-error Intelisense (which allows this) and the actual compiler (which errors) disagree on this.
+				n.toString = undefined;
+				assert.equal(n.toString, undefined);
+			}
+
+			{
+				const n = hydrate(Schema, { toString: undefined });
+				const x = n.toString;
+				assert.equal(x, undefined);
+			}
+		});
+
+		it("optional incompatible shadowing", () => {
+			class Schema extends schemaFactory.object("x", {
+				foo: schemaFactory.optional(schemaFactory.number),
+			}) {
+				// @ts-expect-error incompatible shadowed field errors.
+				public foo(): void {}
+			}
+		});
+
+		it("optional custom shadowing", () => {
+			class Schema extends schemaFactory.object("x", {
+				foo: schemaFactory.optional(schemaFactory.number),
+			}) {
+				// Since fields are own properties, we expect inherited properties (like this) to be shadowed by fields.
+				// However in TypeScript they work like inherited properties, so the types don't make the runtime behavior.
+				// eslint-disable-next-line @typescript-eslint/class-literal-property-style
+				public override get foo(): 5 {
+					return 5;
+				}
+			}
+			function typeTest() {
+				const n = hydrate(Schema, { foo: 1 });
+				assert.equal(n.foo, 1);
+				// @ts-expect-error TypeScript typing does not understand that fields are own properties and thus shadow the getter here.
+				n.foo = undefined;
+			}
+
+			function typeTest2() {
+				const n = hydrate(Schema, { foo: undefined });
+				const x = n.foo;
+				// TypeScript is typing the "foo" field based oin the getter not the field, which does not match runtime behavior.
+				type check_ = requireAssignableTo<typeof x, 5>;
+			}
+
+			assert.throws(
+				() => new Schema({ foo: undefined }),
+				(e: Error) => validateAssertionError(e, /this shadowing will not work/),
+			);
+		});
+
+		it("constructor", () => {
+			// constructor is a special case, since one is built in on the derived type.
+			class Schema extends schemaFactory.object("x", {
+				constructor: schemaFactory.number,
+			}) {}
+
+			const n = hydrate(Schema, { constructor: 5 });
+
+			const x = n.constructor;
+			type check_ = requireAssignableTo<typeof x, number>;
+			assert.equal(x, 5);
+		});
+	});
```

Comment:
One sort of surprising thing I noticed when playing around with classes is that if you do:

```ts
class C extends schemaFactory.object("object", {
	foo: schemaFactory.number,
}) {
	public foo = 3;
}
```

Then it will will actually initialize the "foo" field to 3 whenever an "object" node is created. This happens because the `foo = 3` initializer runs after the class instance has been created, or at least that's what appears to be happening. I'm not sure if that's something we want to advertise, but it does seem kind of neat. For example, it could be a way for users to give defaults to optional fields in the absence of a more powerful defaulting feature. Anyways, if we wanted to add a test that that sort of thing works, I feel like it would belong here, although not necessarily in this PR.

Reply:
Creepy; That will also run when we programatically create an instance from an existing flex tree node, possibly mutating the document when we lazy load a simple tree node. I think thats more of a bug than a feature. Catching it should be possible if we create the node with editing disabled, then enable editing: might be worth doing to ensure such cases error.

Reply:
That's a good point. If the flex node already exists, then we would want any kind of "initializer" to already have run, or not to run at all - but we definitely don't want it to run only when the simple-tree node is created.